runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -30,44 +30,198 @@
 #include "chpltypes.h"
 #include "error.h"
 
-void chpl_mem_layerInit(void) {
-  void* start;
-  size_t size;
 
-  chpl_comm_desired_shared_heap(&start, &size);
+//TODO move the chunk replacement code into a separate file
 
+// TODO put these into a struct?
+void*  heap_base = NULL;
+size_t heap_size = 0;
+size_t cur_heap_offset = 0;
+
+// *** Chunk hook replacements *** //
+// See http://www.canonware.com/download/jemalloc/jemalloc-latest/doc/jemalloc.html#arena.i.chunk_hooks
+
+// Our chunk replacement hook for allocations. Grab memory out of the shared
+// heap and give it to jemalloc.
+static void* chunk_alloc(void *chunk, size_t size, size_t alignment, bool *zero, bool *commit, unsigned arena_ind) {
+
+  void* p = NULL;
+  size_t p_offset;
+
+  // TODO cleanup some of the pointer arithmetic. The current code was
+  // originally copied out of the tcmalloc shim, and could use a little cleanup
+
+  // compute our current pointer into the shared heap based on the current
+  // offset aligned to "alignment"
+  p = (void*)((((uintptr_t) heap_base + cur_heap_offset + alignment - 1) / alignment) * alignment);
+
+  // """
+  // If chunk is not NULL, the returned pointer must be chunk on success or
+  // NULL on error
+  // """
+  if (chunk && chunk != p) {
+    return NULL;
+  }
+
+  p_offset = (uintptr_t)p - (uintptr_t)(heap_base);
+
+  size = ((size + alignment - 1) / alignment) * alignment;
+
+  // If we're out of space on the heap, return NULL
+  if (size > heap_size - p_offset) {
+    return NULL;
+  }
+
+  // Update the current pointer, now that we've past any early returns.
+  cur_heap_offset = p_offset + size;
+
+  // Zero if we were told to
+  if (*zero) {
```

Comment:
Ah, okay, sounds good.  I did some more thinking about NUMA affinity while reviewing the rest of it and I believe the memory locality will actually have to be set prior to this anyway, for a shared heap.  So the part of my comment about that is moot anyway!